File: examples/elixir/lib/worker.ex
# Copyright 2026 Chengxi Luo
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

defmodule Worker do
  @moduledoc """
  Elixir example worker.

  This module is the whole worker: it names the types it can handle, one
  serializer/deserializer pair per format. Everything else lives in the modules
  beside it — those stand in for the types an application already owns, each
  carrying its own schema binding and byte layout.

  telemetry is the one type this worker does not register, and it is not
  coming: the BEAM has no NaN and no infinity, and its float cases carry both.
  It is reported to the runner as SKIPPED and declared in
  examples/cases/expected_skips/elixir.yaml.
  """

  def main(_args) do
    WorkerLib.run_suite(%{
      "customer" => %WorkerLib.Type{
        model: CustomerRecord,
        formats: %{
          "binary" => {&CustomerRecord.marshal/1, &CustomerRecord.unmarshal/1},
          "json" => {&CustomerRecord.to_json/1, &CustomerRecord.from_json/1}
        }
      },
      "ledger" => binary(LedgerEntry),
      "order" => binary(OrderRecord),
      "signals" => binary(SignalCapture),
      "notification" => binary(NotificationRecord)
    })
  end

  # A model that carries its own byte layout, under the one format it speaks.
  # Registered as a %Type{}, so serify converts field map <-> model around
  # marshal/unmarshal and neither of them ever sees a field map.
  defp binary(model) do
    %WorkerLib.Type{
      model: model,
      formats: %{"binary" => {&model.marshal/1, &model.unmarshal/1}}
    }
  end
end

File: examples/cpp/worker.cpp
// Copyright 2026 Chengxi Luo
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// C++ example worker.
//
// This file is the whole worker: it names the types it can handle, one
// serializer/deserializer pair per format. Everything else lives in the headers
// beside it — those stand in for the types an application already owns, each
// carrying its own schema binding and byte layout.
//
// Types this worker does not register (customer, order, telemetry) are reported
// to the runner as SKIPPED.

#include "serify.hpp"

#include "ledger.hpp"
#include "notification.hpp"
#include "signals.hpp"

#include <cstdint>
#include <vector>

using namespace serify;

// One (serializer, deserializer) pair for a model that carries its own byte
// layout: FieldMap in, model, bytes out — and back.
template <typename Model, std::vector<uint8_t> (*Marshal)(const Model&),
          Model (*Unmarshal)(const std::vector<uint8_t>&)>
static FormatPair binary() {
    return FormatPair{
        [](const FieldMap& fm) {
            Model m{};
            from_field_map(fm, m);
            return Marshal(m);
        },
        [](const std::vector<uint8_t>& data) { return to_field_map(Unmarshal(data)); },
    };
}

int main() {
    SuiteMap suite;
    suite["ledger"]["binary"] = binary<LedgerEntry, ledger_marshal, ledger_unmarshal>();
    suite["signals"]["binary"] = binary<SignalCapture, signals_marshal, signals_unmarshal>();
    suite["notification"]["binary"] =
        binary<NotificationRecord, notification_marshal, notification_unmarshal>();
    run_suite(suite);
    return 0;
}
